Improve performance of CheckSuspiciousContent

iSazonov opened the pull request:
<!-- Anything that looks like this is a comment and can't be seen after the Pull Request is created. -->

# PR Summary

By commit:
1. Use MIN_LEN and MAX_LEN constants for better readability.
2. Skip hash searches for length < MIN_LEN.
    - Before we searched for strings with lengths 1, 2, and 3 but shortest pattern (`Emit`) has length 4.
4. Use Span<> instead of array
    - Exclude array allocations 
    - Exclude boundary checks in loops
6. Replace multiplication with more efficient operations
    - Multiplying a value by 31 is multiplying by 32 minus one value. Then change the multiplication to a shift. (The compiler knows how to replace multiplication by a power of two with a shift, but we do this explicitly to be safe.)
7. Improve to-lower-case
    - before we do 4-8 operations, 6 in most cases
    - after we do 3-5 operations, 3 in most cases

Results for `Match("aaaaaaaEmitaaaaaaa")`:
``` ini

BenchmarkDotNet=v0.13.2, OS=Windows 10 (10.0.19044.2251/21H2/November2021Update)
Intel Core i5-2410M CPU 2.30GHz (Sandy Bridge), 1 CPU, 4 logical and 2 physical cores
.NET SDK=7.0.100
  [Host]     : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX
  DefaultJob : .NET 7.0.0 (7.0.22.51805), X64 RyuJIT AVX


```
| Method |               text |       Mean | Ratio | Code Size | Allocated | Alloc Ratio |
|------- |------------------- |-----------:|------:|----------:|----------:|------------:|
|   Orig | aaaaaaaEmitaaaaaaa |   423.8 ns |  1.00 |   4,445 B |     144 B |        1.00 |
|   Fast | aaaaaaaEmitaaaaaaa |   266.0 ns |  0.63 |   4,493 B |         - |        0.00 |

## PR Context

<!-- Provide a little reasoning as to why this Pull Request helps and why you have opened it. -->

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): <!-- Experimental feature name(s) here -->
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: <!-- Number/link of that issue here -->
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: <!-- Number/link of that issue here -->
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed: <!-- Number/link of that issue here -->

iSazonov committed:
Add MIN_LEN and MAX_LEN constants

iSazonov committed:
Skip all before shortest suspicious string

iSazonov committed:
Exclude allocations replacing array with Span

iSazonov committed:
Replace multiplication with more efficient operations

iSazonov committed:
Improve to-lower-case

SteveL-MSFT commented:
`LookUpHash()` uses a large switch statement, would it be faster to change to a dictionary?

iSazonov commented:
> `LookUpHash()` uses a large switch statement, would it be faster to change to a dictionary?

I tested this and found the switch statement is more faster (looking asm code I guess C# compiler tries to do a binary search).

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -2012,28 +2019,26 @@ private static string CheckForMatches(uint[] runningHash, int upTo)
             /// <returns>The string matching the hash, or null.</returns>
             public static string Match(string text)
```
When reviewing the changes, I found the algorithm in `Match` has flaws: when the text starts with a suspicious text defined in `LookupHash(uint h)`, it cannot find the text.

#### dotnet application to reproduce the issue

```c#
      static void Main(string[] args)
      {
          string match = SuspiciousContentChecker.Match("Add-Type");
          if (match is null)
          {
              Console.WriteLine("match not found.");
          }
          else
          {
              Console.WriteLine($"match found: {match}");
          }
      }

     // Copy the SuspiciousContentChecker class here.
     // ....
```

Results from running the application:

![image](https://user-images.githubusercontent.com/127450/204669157-fed84338-b03a-4ff6-9e4b-7335a4ef012b.png)

To make the `Match` work, at least 1 more character needs to be present before `Add-Type`. Change it to `SuspiciousContentChecker.Match(" Add-Type");`, then it's able to find the `Add-Type` text:

![image](https://user-images.githubusercontent.com/127450/204669585-324c0636-f0f7-40ee-9915-6c129369cf8e.png)


----

The `Match` call is used in [CheckSuspiciousContent(Ast scriptBlockAst)](https://github.com/PowerShell/PowerShell/blob/d03a5553c237c39dcbe7fac0c1d39eef9f140ec6/src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs#LL1775-L1779C14):
```
            var foundSignature = SuspiciousContentChecker.Match(scriptBlockAst.Extent.Text);
            if (foundSignature != null)
            {
                return foundSignature;
            }
```
And of course, we can create a script block starts with `Add-Type`, for example: 

![image](https://user-images.githubusercontent.com/127450/204670373-412f10f8-8312-489d-91a8-a0b434a2fb74.png)

daxian-dbw replied:
I think the fix is: instead of
```
for (int j = Math.Min(i, runningHash.Length) - 1; j > 0; j--)
```
we should remove the `- 1` part:
```
for (int j = Math.Min(i, runningHash.Length); j > 0; j--)
```

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -2026,12 +2026,10 @@ public static string Match(string text)
-                        h |= 0x20; // ToLower
-                    }
-                    else if (!((h >= 'a' && h <= 'z') || h == '-'))
+                    h |= 0x20;
```
Will blindly doing this cause any problem? `h` could be any character, not just ASCII ones.
The original code is way more readable than this, and I double if the perf gain is worth sacrificing the readability.

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -1988,7 +1988,8 @@ private static string CheckForMatches(uint[] runningHash, int upTo)
 
-                for (var i = 0; i < upTo; i++)
+                // Skip all before shortest suspicious string. Today it is 'Emit' so start with i=3.
+                for (var i = MIN_LEN - 1; i < upTo; i++)
```
By looking at the hashing algorithm used in `Match`, we should just only check `upTo - 1` here, no need to do the loop if the passed-in `upTo` is not 0. The loop is needed only when the passed in `upTo` is 0.

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -1972,15 +1976,18 @@ private static string LookupHash(uint h)
-                for (var i = 0; i < upTo; i++)
+                runningHash = runningHash.Slice(0, upTo);
+
+                // Skip all before shortest suspicious string. Today it is 'Emit' so start with i=3.
```
```suggestion
                // Starts from where the hash for the shortest suspicious string would be stored.
```

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -2012,28 +2019,26 @@ private static string CheckForMatches(uint[] runningHash, int upTo)
-                        h |= 0x20; // ToLower
-                    }
-                    else if (!((h >= 'a' && h <= 'z') || h == '-'))
+                    h |= 0x20;
+                    if (!((uint)(h - 'a') <= 'z' - 'a' || h == '-'))
```
Will blindly doing `h |= 0x20` cause any problem? `h` could be any character, not just ASCII ones.
The original code is way more readable than this, and I double if the perf gain is worth sacrificing the readability.

iSazonov replied:
It is not problem because we reject all except A-Z, a-z, and `-`.
Perf win of the change is ~10% because:
- old code optimized for upper case chars
- new code does less operations for any chars.

daxian-dbw replied:
We are talking about improvements in nanosecond level. I don't think it's the bottleneck to the overall script execution, and because of that, I prefer to keep the original code for readability reason.

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -2012,28 +2019,26 @@ private static string CheckForMatches(uint[] runningHash, int upTo)
                     }
 
-                    for (int j = Math.Min(i, runningHash.Length) - 1; j > 0; j--)
+                    Span<uint> rh = runningHash.Slice(0, Math.Min(i, runningHash.Length));
```
Why doing a `Slice` here? Can't we use `runningHash` directly in the loop below?

iSazonov replied:
The change allows compiler to remove boundary checks.

daxian-dbw replied:
Please don't use `Slice` here, so as to make a separate fix easier. Just replacing the array with span would be a great change.

iSazonov replied:
I removed the Slice but this stopping compiler from removing boundary checks in the loop.

iSazonov replied:
We can have the Slice with comment:
```c#
// We need cut unused tail. Since 'i' is current position, the actual length is 'i + 1'. Ex., if i = 0 (first char in the string) the span length is 1, if i = 1 (second char) the span length is 2, and so on.
Span<uint> rh = runningHash.Slice(0, Math.Min(i + 1, runningHash.Length));
```

daxian-dbw replied:
Given that `rh[j - 1]` is used in the loop, I doubt the boundary check on that can be eliminated too.

iSazonov replied:
Yes, this works for `rh[j]` only. It is easy to exclude for `rh[j - 1]` too. Really we have no need to read previous value since we already have it in `v` variable.
I was thinking about vectorization, but I assume that you will be against the complication of the code. :-)

Since we have #18693 with tests can we return span?

daxian-dbw replied:
Nope, let's not using `Slice` here, just to keep the code simple and readable.
If you want, you can eliminate the call to `Math.Min` by doing `j = val1 > val2 ? val1 : val2` directly before the loop.

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -2012,28 +2019,26 @@ private static string CheckForMatches(uint[] runningHash, int upTo)
 
-                    for (int j = Math.Min(i, runningHash.Length) - 1; j > 0; j--)
+                    Span<uint> rh = runningHash.Slice(0, Math.Min(i, runningHash.Length));
+                    for (int j = rh.Length - 1; j > 0; j--)
```
According to https://github.com/PowerShell/PowerShell/pull/18665/files#r1035381609, the `- 1` part should be removed.

```suggestion
                    for (int j = rh.Length; j > 0; j--)
```

iSazonov replied:
Do you mean:
```suggestion
                    for (int j = rh.Length - 1; j >= 0; j--)
```

iSazonov replied:
`rh[j]` will throw.

daxian-dbw replied:
Yeah, it will throw with your changes. With the original code, remove the `-1` part will fix the issue.

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -2043,15 +2048,19 @@ public static string Match(string text)
                         //    This overwrites the previous iteration, hence we go from longest to shortest.
                         //
-                        // LCG comes from a trivial (bad) random number generator,
+                        // Original code:
+                        // rh[j] = 31 * rh[j - 1] + h;
+                        // 31 comes from a trivial (bad) random number generator,
```
```suggestion
                        // The original code is:   rh[j] = 31 * rh[j - 1] + h;
                        // The number '31' comes from a trivial (bad) random number generator,
```

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -2043,15 +2048,19 @@ public static string Match(string text)
-                        runningHash[j] = LCG * runningHash[j - 1] + h;
+                        //
+                        // Then replace multiplication with more efficient operations.
+                        rh[j] = (rh[j - 1] << 5) - rh[j - 1] + h;
```
Given that you have gone so far, maybe do the following change to avoid additional boundary checks?

```suggestion
                        uint v = rh[j - 1];
                        rh[j] = (v << 5) - v + h;
```

iSazonov replied:
We can if you want but compiler does the optimization.

daxian-dbw replied:
> The compiler knows how to replace multiplication by a power of two with a shift, but we do this explicitly to be safe

This is from your PR description. If so, then I'd prefer to not do anything, just rely on the compiler to do the right thing.
It's a micro-optimization anyway, which is likely won't make a difference in overall script execution performance.

iSazonov committed:
Apply suggestions from code review

Co-authored-by: Dongbo Wang <dongbow@microsoft.com>

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -2012,28 +2019,26 @@ private static string CheckForMatches(uint[] runningHash, int upTo)
 
-                    for (int j = Math.Min(i, runningHash.Length) - 1; j > 0; j--)
+                    Span<uint> rh = runningHash.Slice(0, Math.Min(i, runningHash.Length));
+                    for (int j = rh.Length - 1; j > 0; j--)
```
Just a note here that a separate fix is tracked internally.

daxian-dbw replied:
https://github.com/PowerShell/PowerShell/pull/18693

iSazonov committed:
Remove Span

daxian-dbw commented on src/System.Management.Automation/engine/runtime/CompiledScriptBlock.cs:
```diff
@@ -1972,15 +1976,18 @@ private static string LookupHash(uint h)
+                runningHash = runningHash.Slice(0, upTo);
+
+                // Starts from where the hash for the shortest suspicious string would be stored.
+                for (var i = MIN_LEN - 1; i < runningHash.Length; i++)
```
Given this is not a `for (i = 0; i < span.Length; i++)` pattern, are you sure the boundary check will be eliminated?

iSazonov replied:
Yes, this works. https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8aHunHiUGAWQAUASg5ceUyADtcGBgEsZCwQEkAcgH0AMgFENDALwMUAbklTuAN2xQGAC2MMZMAO4MwDuwG1yABn8AXQtqKx4AM2gGYVt7RWd1bX1DOAZyMyUGAB5HADodGBkAcwwHTMUAakrRSytOMPCrYnIATmEHH0Ug0VCmhgBfOqHqAaA==

daxian-dbw replied:
That doesn't reflect the actual situation of this method. [This](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYQYG8GuGbunHiKBgDkAFAGUADtgB2AHjAALbFAB8DRWgYBLGRgYBXSQBUIASl7d2lvl0gzc+3foCyASWEB9ADIBRYQwAvAwoANw2topBGgB04gA22mAwogAMWkamZuHUttwAZtAMogBuKjrR7l5+AXAM5KEVcrHeMDIA5hiKjdoA1L0WuXkcEcNM5ACcoooA2toAutmj3AC+EWtDXBH0TEIAChLS8koq6po6eoYm5hHWm7b2jheuHj7+0WHLXIVQxWW/2kqrxqDDqDSaV1MPX6gzGdzGfGIk2mc0WOTGGz4GxWQA==) does.

Having `Slice` here moves the boundary check from within the loop to outside the loop, so it is better. Also, it doesn't affect readability, so it's good to have this change.

iSazonov committed:
Revert "Remove Span"

This reverts commit c5ef64148cce901daf06476d95b07b12b286238b.

iSazonov committed:
Use Span

iSazonov committed:
Revert loop and check length

iSazonov committed:
Fix length

iSazonov commented:
@daxian-dbw I changed the direction of the loop so that all the boundary checks are removed. Previously, I removed all allocations and made a quick character filter. So now this is the fastest version.
In addition, I added result length check, which should noticeably reduce the number of collisions in real-world scenarios. If Travis sees it in his tests, maybe you (team) can approve this PR. Otherwise just close it.

daxian-dbw commented:
@iSazonov @TravisEz13's comments were not addressed. Also, I doubt if Travis still has his original tests available, but will let Travis to confirm.

iSazonov committed:
Replace LCG with 31

iSazonov commented:
LOG const remove at all.
